lib/internal/errors.js

```diff
@@ -371,42 +372,101 @@ function makeSystemErrorWithCode(key) {
-    // Reset the limit and setting the name property.
-    if (isErrorStackTraceLimitWritable()) Error.stackTraceLimit = limit;
-    const message = getMessage(key, args, error);
-    ObjectDefineProperties(error, {
-      [kIsNodeError]: {
-        __proto__: null,
-        value: true,
-        enumerable: false,
-        writable: false,
-        configurable: true,
-      },
-      message: {
-        __proto__: null,
-        value: message,
-        enumerable: false,
-        writable: true,
-        configurable: true,
-      },
-      toString: {
-        __proto__: null,
-        value() {
+  const defaultMessage = messages.get(key);
+  const expectedLength = typeof defaultMessage !== 'string' ? -1 : getExpectedArgumentLength(defaultMessage);
+
+  switch (expectedLength) {
+    case 0: {
+      class NodeError extends Base {
+        code = key;
+
+        constructor(...args) {
+          assert(
+            args.length === 0,
+            `Code: ${key}; The provided arguments length (${args.length}) does not ` +
+              `match the required ones (${expectedLength}).`,
+          );
+          super(defaultMessage);
+        }
+
+        // This is a workaround for wpt tests that expect that the error
+        // constructor has a `name` property of the base class.
+        get ['constructor']() {
+          return Base;
+        }
+
+        get [kIsNodeError]() {
+          return true;
+        }
+
+        toString() {
           return `${this.name} [${key}]: ${this.message}`;
-        },
-        enumerable: false,
-        writable: true,
-        configurable: true,
-      },
-    });
-    captureLargerStackTrace(error);
-    error.code = key;
-    return error;
-  };
+        }
+      }
+      return NodeError;
+    }
+    case -1: {
+      class NodeError extends Base {
+        code = key;
+
+        constructor(...args) {
+          super();
+          ObjectDefineProperty(this, 'message', {
+            __proto__: null,
+            value: getMessage(key, args, this),
+            enumerable: false,
+            writable: true,
+            configurable: true,
+          });
+        }
+
+        // This is a workaround for wpt tests that expect that the error
+        // constructor has a `name` property of the base class.
+        get ['constructor']() {
+          return Base;
+        }
+
+        get [kIsNodeError]() {
+          return true;
+        }
+
+        toString() {
+          return `${this.name} [${key}]: ${this.message}`;
+        }
+      }
+      return NodeError;
+    }
+    default: {
+
+      class NodeError extends Base {
+        code = key;
+
+        constructor(...args) {
+          assert(
+            args.length === expectedLength,
+            `Code: ${key}; The provided arguments length (${args.length}) does not ` +
+              `match the required ones (${expectedLength}).`,
+          );
+          super(ReflectApply(lazyInternalUtilInspect().format, null, args));
```

Comment:
Do we really need ReflectApply here? It seems fine to just call `lazyInternalUtilInspect().format(...args)` here, that should probably be faster.

Reply:
Yeah, but what was the reason first, that it used ReflectApply?

Reply:
`…args` relies on the publicly mutable `Array.prototype[Symbol.iterator]`

Reply:
So ReflectApply is needed?

Reply:
I see, that's a bummer, but I guess this is one of the paths where we do need to strengthen against prototype pollution. Would be useful to leave a comment though.